File: sdk/typescript/service.ts
import type {Http2Server} from 'node:http2';
import {createServer} from 'node:http2';
import {createClient, type Client, type ConnectRouter} from '@connectrpc/connect';
import {connectNodeAdapter, createGrpcTransport} from '@connectrpc/connect-node';
import {create} from '@bufbuild/protobuf';
import {timestampDate} from '@bufbuild/protobuf/wkt';
import {type ExecuteRequest, type ExecuteResponse, Outcome as WireOutcome, ToolExecutionService} from './gen/flory/gateway/v1/execution_pb.js';
import {HealthReportSchema, InstanceInfoSchema, RegistryService, ServingStatus, type ToolStatus, ToolState} from './gen/flory/gateway/v1/registry_pb.js';
import type {ToolContract} from './gen/flory/gateway/v1/tool_contract_pb.js';
import {Health, HealthCheckResponse_ServingStatus} from './gen-health/grpc/health/v1/health_pb.js';
import {buildContract, type Contract, contractViolation} from './contract.js';

/** What a handler reports about one attempt, in the executors' own vocabulary. */
export type Outcome = 'succeeded' | 'retryable-failure' | 'permanent-failure' | 'unknown';

/** One attempt delivered to a handler. */
export interface Call {
    runId: string;
    vertexId: string;
    scopeId: string;
    attemptNo: number;
    toolId: string;
    toolVersion: string;
    idempotencyKey: string;
    /** Arguments already validated by the gateway against the frozen input schema. */
    args: Record<string, unknown>;
    deadline?: Date;
}

/** A handler's answer. */
export interface Result {
    outcome: Outcome;
    value?: unknown;
    error?: string;
}

/**
 * Executes one attempt of one tool.
 *
 * It is called at most once per attempt: neither the SDK nor the gateway will call it again on its own. If the same
 * idempotency key arrives twice, that is an executor's deliberate retry, and the handler must be idempotent in
 * exactly the way its contract declared.
 */
export type Handler = (call: Call) => Promise<Result> | Result;

/** A declared contract paired with the handler that serves it. */
export interface Tool {
    contract: Contract;
    handler: Handler;
}

/** Describes one tool-service instance. */
export interface ServiceConfig {
    /** Unique per process. A restarted process may reuse it; re-registering an identical contract is idempotent. */
    instanceId: string;
    /** The logical route this service serves. Several instances share it, which is what lets the gateway spread attempts. */
    routeId: string;
    /** The address the gateway will reach this instance on, which is not always the listen address. */
    target: string;
    /** The gateway's registration surface, as a base URL. */
    gatewayUrl: string;
    leaseTtlMs?: number;
    heartbeatIntervalMs?: number;
    log?: (message: string, detail?: unknown) => void;
}

/** Keeps several beats inside one lease, so a single lost heartbeat does not withdraw a healthy instance. */
export const DEFAULT_LEASE_TTL_MS = 30_000;
/** The default heartbeat cadence. */
export const DEFAULT_HEARTBEAT_INTERVAL_MS = 5_000;

const OUTCOMES: Record<Outcome, WireOutcome> = {
    succeeded: WireOutcome.SUCCEEDED,
    'retryable-failure': WireOutcome.RETRYABLE_FAILURE,
    'permanent-failure': WireOutcome.PERMANENT_FAILURE,
    unknown: WireOutcome.UNKNOWN,
};

/** One registering, heartbeating tool service. */
export class ToolService {
    private readonly tools = new Map<string, Tool>();
    private readonly contracts = new Map<string, ToolContract>();
    private readonly registry: Client<typeof RegistryService>;
    private readonly log: (message: string, detail?: unknown) => void;
    private server?: Http2Server;
    private heartbeat?: ReturnType<typeof setInterval>;
    private serving = false;

    constructor(private readonly config: ServiceConfig) {
        if (!config.instanceId || !config.routeId || !config.target) throw new Error('sdk: instanceId, routeId, and target are all required');
        this.log = config.log ?? (() => {});
        this.registry = createClient(RegistryService, createGrpcTransport({baseUrl: config.gatewayUrl}));
    }

    /**
     * Adds tools this service offers.
     *
     * Contracts are validated here, at startup, before any network call: a service that declares a saga with no
     * compensator should fail while someone is watching it start.
     */
    declare(...tools: readonly Tool[]): void {
        for (const tool of tools) {
            const violation = contractViolation(tool.contract);
            if (violation) throw new Error(`sdk: ${tool.contract.toolId} is not registrable: ${violation}`);
            if (this.tools.has(tool.contract.toolId)) throw new Error(`sdk: ${tool.contract.toolId} is declared twice`);
            this.tools.set(tool.contract.toolId, tool);
            this.contracts.set(tool.contract.toolId, buildContract(tool.contract, this.config.routeId));
        }
    }

    /** Starts the gRPC surface the gateway dispatches to, and its health check. */
    async listen(port: number, host = '127.0.0.1'): Promise<number> {
        const routes = (router: ConnectRouter): void => {
            router.service(ToolExecutionService, {execute: (request) => this.execute(request)});
            // The gateway probes this itself rather than trusting the heartbeat report, so it has to answer even
            // while the service considers itself unready.
            router.service(Health, {
                check: () => ({status: this.serving ? HealthCheckResponse_ServingStatus.SERVING : HealthCheckResponse_ServingStatus.NOT_SERVING}),
            });
        };
        // gRPC mandates HTTP/2, so this is an h2c server rather than the HTTP/1.1
        // one a Connect-only service could use. A gRPC client -- the Go gateway --
        // cannot talk to an HTTP/1.1 listener at all.
        this.server = createServer(connectNodeAdapter({routes, grpc: true}));
        await new Promise<void>((resolve) => this.server!.listen(port, host, resolve));
        const address = this.server.address();
        return typeof address === 'object' && address ? address.port : port;
    }

    /**
     * Publishes this instance's own readiness.
     *
     * It drives both signals the gateway consults: the health check it probes, and the report carried on each
     * heartbeat. A service that knows it cannot work says so here and stops receiving attempts, without deregistering.
     */
    setServing(serving: boolean): void {
        this.serving = serving;
    }

    /** Declares this instance and its contracts to the gateway. */
    async register(): Promise<readonly ToolStatus[]> {
        const response = await this.registry.register({
            instance: create(InstanceInfoSchema, {
                instanceId: this.config.instanceId,
                routeId: this.config.routeId,
                target: this.config.target,
                leaseTtlMs: this.config.leaseTtlMs ?? DEFAULT_LEASE_TTL_MS,
            }),
            tools: this.declaredContracts(),
            health: this.healthReport(),
        });
        for (const status of response.statuses) {
            if (status.state === ToolState.REJECTED) this.log('tool registration rejected', {tool: status.toolId, code: status.code, detail: status.detail});
        }
        return response.statuses;
    }

    /**
     * Keeps the registration alive until stop() is called.
     *
     * Heartbeats are retried on transport failure, unlike a tool call: a heartbeat is idempotent control-plane traffic
     * with no side effect, so repeating one is free, while repeating a tool call could duplicate a business effect.
     *
     * When the gateway reports this instance unknown, this registers again. That is the whole recovery path for a
     * gateway that deliberately keeps no durable registry: it restarts empty, and every live service tells it what it
     * serves again.
     */
    start(): void {
        const interval = this.config.heartbeatIntervalMs ?? DEFAULT_HEARTBEAT_INTERVAL_MS;
        this.heartbeat = setInterval(() => {
            void (async () => {
                try {
                    const response = await this.registry.heartbeat({instanceId: this.config.instanceId, health: this.healthReport()});
                    if (!response.knownInstance) {
                        this.log('gateway does not know this instance; registering again');
                        await this.register();
                    }
                } catch (error: unknown) {
                    this.log('heartbeat failed', error);
                }
            })();
        }, interval);
    }

    /**
     * Deregisters and stops serving.
     *
     * Deregistration withdraws this instance's route. It does not withdraw the contracts: the published view and its
     * digest are unchanged, so a subgraph frozen against them still resolves, and calls fail as unroutable rather
     * than as unknown.
     */
    async stop(): Promise<void> {
        this.setServing(false);
        if (this.heartbeat) clearInterval(this.heartbeat);
        try {
            await this.registry.deregister({instanceId: this.config.instanceId});
        } catch (error: unknown) {
            this.log('deregistration failed', error);
        }
        if (this.server) await new Promise<void>((resolve) => this.server!.close(() => resolve()));
    }

    private declaredContracts(): ToolContract[] {
        // Stable order, so two runs of the same service send byte-identical registrations.
        return [...this.contracts.keys()].sort().map((name) => this.contracts.get(name)!);
    }

    private healthReport() {
        return create(HealthReportSchema, {status: this.serving ? ServingStatus.SERVING : ServingStatus.NOT_SERVING});
    }

    private async execute(request: ExecuteRequest): Promise<ExecuteResponse> {
        const tool = this.tools.get(request.toolId);
        if (!tool) {
            return {outcome: WireOutcome.PERMANENT_FAILURE, result: '', error: `${this.config.instanceId} does not serve ${request.toolId}`} as ExecuteResponse;
        }
        let args: Record<string, unknown> = {};
        if (request.arguments) args = JSON.parse(request.arguments) as Record<string, unknown>;
        const result = await tool.handler({
            runId: request.runId,
            vertexId: request.vertexId,
            scopeId: request.scopeId,
            attemptNo: request.attemptNo,
            toolId: request.toolId,
            toolVersion: request.toolVersion,
            idempotencyKey: request.idempotencyKey,
            args,
            deadline: request.deadline ? timestampDate(request.deadline) : undefined,
        });
        const outcome = OUTCOMES[result.outcome];
        if (outcome === undefined) {
            // A handler that returned no outcome has told us nothing about whether its effect happened, and unknown is
            // exactly that statement.
            return {outcome: WireOutcome.UNKNOWN, result: '', error: `handler returned an unknown outcome ${String(result.outcome)}`} as ExecuteResponse;
        }
        return {outcome, result: result.value === undefined ? '' : JSON.stringify(result.value), error: result.error ?? ''} as ExecuteResponse;
    }
}

File: gatewayd/sdk/service.go
package sdk

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"net"
	"sort"
	"sync"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/health"
	"google.golang.org/grpc/health/grpc_health_v1"
	"google.golang.org/protobuf/types/known/timestamppb"

	gatewayv1 "github.com/linuxb/flory-ai/gatewayd/internal/pb/flory/gateway/v1"
)

// Outcome is what a handler reports about one attempt.
type Outcome string

// The outcome vocabulary, identical to the one the executors reason with.
const (
	// OutcomeSucceeded means the operation completed.
	OutcomeSucceeded Outcome = "succeeded"
	// OutcomeRetryableFailure means the same idempotent operation may be
	// attempted again -- by the executor that owns the retry decision.
	OutcomeRetryableFailure Outcome = "retryable-failure"
	// OutcomePermanentFailure means retry cannot resolve the failure.
	OutcomePermanentFailure Outcome = "permanent-failure"
	// OutcomeUnknown means the service cannot say whether its effect happened.
	OutcomeUnknown Outcome = "unknown"
)

// Call is one attempt delivered to a handler.
type Call struct {
	RunID          string
	VertexID       string
	ScopeID        string
	AttemptNo      uint32
	ToolID         string
	ToolVersion    string
	IdempotencyKey string
	Arguments      json.RawMessage
	Deadline       time.Time
}

// Arg decodes the call arguments into value.
func (call Call) Arg(value any) error {
	return json.Unmarshal(call.Arguments, value)
}

// Result is a handler's answer.
type Result struct {
	Outcome Outcome
	// Value is marshalled to JSON as the tool's result.
	Value any
	Error string
}

// Handler executes one attempt of one tool.
//
// It is called at most once per attempt: neither the SDK nor the gateway will
// call it again on its own. If the same idempotency key arrives twice, that is
// an executor's deliberate retry, and the handler is expected to be idempotent
// in exactly the way its contract declared.
type Handler func(ctx context.Context, call Call) Result

// Tool pairs a declared contract with the handler that serves it.
type Tool struct {
	Contract Contract
	Handler  Handler
}

// Config describes one tool-service instance.
type Config struct {
	// InstanceID must be unique per process. A restarted process may reuse it;
	// the gateway treats re-registration of an identical contract as idempotent.
	InstanceID string
	// RouteID is the logical route this service serves. Several instances of one
	// service share it, which is what lets the gateway spread attempts.
	RouteID string
	// Target is the dial address the gateway will reach this instance on. It must
	// be reachable from the gateway, which is not always the listen address.
	Target string
	// GatewayAddress is the gateway's registration surface, "host:port".
	GatewayAddress    string
	LeaseTTL          time.Duration
	HeartbeatInterval time.Duration
	Logger            *slog.Logger
}

// DefaultLeaseTTL and DefaultHeartbeatInterval keep several beats inside one
// lease, so a single lost heartbeat does not withdraw a healthy instance.
const (
	DefaultLeaseTTL          = 30 * time.Second
	DefaultHeartbeatInterval = 5 * time.Second
)

// Service is one registering, heartbeating tool service.
type Service struct {
	gatewayv1.UnimplementedToolExecutionServiceServer
	config    Config
	logger    *slog.Logger
	mutex     sync.RWMutex
	tools     map[string]Tool
	contracts map[string]*gatewayv1.ToolContract

	grpcServer   *grpc.Server
	healthServer *health.Server

	clientMutex sync.Mutex
	connection  *grpc.ClientConn
	client      gatewayv1.RegistryServiceClient
}

// NewService creates a tool service that has not yet declared any tools.
func NewService(config Config) (*Service, error) {
	if config.InstanceID == "" || config.RouteID == "" || config.Target == "" {
		return nil, errors.New("sdk: InstanceID, RouteID, and Target are all required")
	}
	if config.LeaseTTL <= 0 {
		config.LeaseTTL = DefaultLeaseTTL
	}
	if config.HeartbeatInterval <= 0 {
		config.HeartbeatInterval = DefaultHeartbeatInterval
	}
	if config.Logger == nil {
		config.Logger = slog.Default()
	}
	service := &Service{
		config:       config,
		logger:       config.Logger,
		tools:        map[string]Tool{},
		contracts:    map[string]*gatewayv1.ToolContract{},
		grpcServer:   grpc.NewServer(),
		healthServer: health.NewServer(),
	}
	gatewayv1.RegisterToolExecutionServiceServer(service.grpcServer, service)
	grpc_health_v1.RegisterHealthServer(service.grpcServer, service.healthServer)
	service.SetServing(false)
	return service, nil
}

// Declare adds a tool this service offers.
//
// The contract is validated here, at startup, against the gateway's own rules --
// before any network call. A service that declares snapshot compensation or a
// saga with no compensator should fail while someone is watching it start, not
// on a rejection it has to go read a log to find.
func (service *Service) Declare(tools ...Tool) error {
	service.mutex.Lock()
	defer service.mutex.Unlock()
	for _, tool := range tools {
		contract, err := tool.Contract.Build(service.config.RouteID)
		if err != nil {
			return err
		}
		if err := validateContract(contract); err != nil {
			return fmt.Errorf("sdk: %s is not registrable: %w", tool.Contract.ToolID, err)
		}
		if _, declared := service.tools[tool.Contract.ToolID]; declared {
			return fmt.Errorf("sdk: %s is declared twice", tool.Contract.ToolID)
		}
		service.tools[tool.Contract.ToolID] = tool
		service.contracts[tool.Contract.ToolID] = contract
	}
	return nil
}

// Serve starts the gRPC surface on listener and returns when it stops.
func (service *Service) Serve(listener net.Listener) error {
	return service.grpcServer.Serve(listener)
}

// SetServing publishes this instance's own readiness.
//
// It drives both health signals the gateway consults: the gRPC health check it
// probes, and the report carried on each heartbeat. A service that knows it
// cannot work says so here and stops receiving attempts, without deregistering.
func (service *Service) SetServing(serving bool) {
	status := grpc_health_v1.HealthCheckResponse_NOT_SERVING
	if serving {
		status = grpc_health_v1.HealthCheckResponse_SERVING
	}
	service.healthServer.SetServingStatus("", status)
	service.healthServer.SetServingStatus(gatewayv1.ToolExecutionService_ServiceDesc.ServiceName, status)
}

func (service *Service) serving() bool {
	response, err := service.healthServer.Check(context.Background(), &grpc_health_v1.HealthCheckRequest{})
	return err == nil && response.GetStatus() == grpc_health_v1.HealthCheckResponse_SERVING
}

// Execute dispatches one attempt to its handler.
func (service *Service) Execute(ctx context.Context, request *gatewayv1.ExecuteRequest) (*gatewayv1.ExecuteResponse, error) {
	service.mutex.RLock()
	tool, known := service.tools[request.GetToolId()]
	service.mutex.RUnlock()
	if !known {
		return &gatewayv1.ExecuteResponse{
			Outcome: gatewayv1.Outcome_OUTCOME_PERMANENT_FAILURE,
			Error:   fmt.Sprintf("%s does not serve %s", service.config.InstanceID, request.GetToolId()),
		}, nil
	}
	arguments := json.RawMessage(request.GetArguments())
	if len(arguments) == 0 {
		arguments = json.RawMessage(`{}`)
	}
	if deadline := request.GetDeadline(); deadline != nil {
		var cancel context.CancelFunc
		ctx, cancel = context.WithDeadline(ctx, deadline.AsTime())
		defer cancel()
	}
	result := tool.Handler(ctx, Call{
		RunID:          request.GetRunId(),
		VertexID:       request.GetVertexId(),
		ScopeID:        request.GetScopeId(),
		AttemptNo:      request.GetAttemptNo(),
		ToolID:         request.GetToolId(),
		ToolVersion:    request.GetToolVersion(),
		IdempotencyKey: request.GetIdempotencyKey(),
		Arguments:      arguments,
		Deadline:       request.GetDeadline().AsTime(),
	})
	return renderResult(result)
}

var outcomeCodes = map[Outcome]gatewayv1.Outcome{
	OutcomeSucceeded:        gatewayv1.Outcome_OUTCOME_SUCCEEDED,
	OutcomeRetryableFailure: gatewayv1.Outcome_OUTCOME_RETRYABLE_FAILURE,
	OutcomePermanentFailure: gatewayv1.Outcome_OUTCOME_PERMANENT_FAILURE,
	OutcomeUnknown:          gatewayv1.Outcome_OUTCOME_UNKNOWN,
}

func renderResult(result Result) (*gatewayv1.ExecuteResponse, error) {
	code, known := outcomeCodes[result.Outcome]
	if !known {
		// A handler that returned no outcome has told us nothing about whether its
		// effect happened, and unknown is exactly that statement.
		return &gatewayv1.ExecuteResponse{Outcome: gatewayv1.Outcome_OUTCOME_UNKNOWN, Error: fmt.Sprintf("handler returned an unknown outcome %q", result.Outcome)}, nil
	}
	encoded := ""
	if result.Value != nil {
		payload, err := json.Marshal(result.Value)
		if err != nil {
			return &gatewayv1.ExecuteResponse{Outcome: gatewayv1.Outcome_OUTCOME_UNKNOWN, Error: fmt.Sprintf("result is not encodable: %v", err)}, nil
		}
		encoded = string(payload)
	}
	return &gatewayv1.ExecuteResponse{Outcome: code, Result: encoded, Error: result.Error}, nil
}

// Register declares this instance and its contracts to the gateway.
func (service *Service) Register(ctx context.Context) ([]*gatewayv1.ToolStatus, error) {
	client, err := service.registryClient()
	if err != nil {
		return nil, err
	}
	response, err := client.Register(ctx, &gatewayv1.RegisterRequest{
		Instance: service.instanceInfo(),
		Tools:    service.declaredContracts(),
		Health:   service.healthReport(),
	})
	if err != nil {
		return nil, fmt.Errorf("sdk: register: %w", err)
	}
	for _, status := range response.GetStatuses() {
		if status.GetState() == gatewayv1.ToolState_TOOL_STATE_REJECTED {
			service.logger.Error("tool registration rejected", "tool", status.GetToolId(), "code", status.GetCode().String(), "detail", status.GetDetail())
		}
	}
	return response.GetStatuses(), nil
}

// Run keeps the registration alive until the context is cancelled.
//
// Heartbeats are retried on transport failure, unlike a tool call: a heartbeat
// is idempotent control-plane traffic with no side effect, so repeating one is
// free, while repeating a tool call could duplicate a business effect.
//
// When the gateway reports the instance unknown, this re-registers. That is the
// whole recovery path for a gateway that deliberately keeps no durable registry:
// it restarts empty, and every live service tells it what it serves again.
func (service *Service) Run(ctx context.Context) error {
	ticker := time.NewTicker(service.config.HeartbeatInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-ticker.C:
			client, err := service.registryClient()
			if err != nil {
				service.logger.Warn("gateway unreachable", "error", err)
				continue
			}
			response, err := client.Heartbeat(ctx, &gatewayv1.HeartbeatRequest{InstanceId: service.config.InstanceID, Health: service.healthReport()})
			if err != nil {
				service.logger.Warn("heartbeat failed", "error", err)
				continue
			}
			if !response.GetKnownInstance() {
				service.logger.Info("gateway does not know this instance; registering again")
				if _, err := service.Register(ctx); err != nil {
					service.logger.Warn("re-registration failed", "error", err)
				}
			}
		}
	}
}

// Shutdown deregisters and stops serving.
//
// Deregistration withdraws this instance's route. It does not withdraw the
// contracts: the published view and its digest are unchanged, so a subgraph
// frozen against them still resolves, and calls fail as unroutable rather than
// as unknown.
func (service *Service) Shutdown(ctx context.Context) error {
	service.SetServing(false)
	var failure error
	if client, err := service.registryClient(); err == nil {
		if _, err := client.Deregister(ctx, &gatewayv1.DeregisterRequest{InstanceId: service.config.InstanceID}); err != nil {
			failure = fmt.Errorf("sdk: deregister: %w", err)
		}
	}
	service.clientMutex.Lock()
	if service.connection != nil {
		_ = service.connection.Close()
		service.connection, service.client = nil, nil
	}
	service.clientMutex.Unlock()
	service.grpcServer.GracefulStop()
	return failure
}

func (service *Service) instanceInfo() *gatewayv1.InstanceInfo {
	return &gatewayv1.InstanceInfo{
		InstanceId: service.config.InstanceID,
		RouteId:    service.config.RouteID,
		Target:     service.config.Target,
		LeaseTtlMs: uint32(service.config.LeaseTTL.Milliseconds()),
	}
}

func (service *Service) healthReport() *gatewayv1.HealthReport {
	status := gatewayv1.ServingStatus_SERVING_STATUS_NOT_SERVING
	if service.serving() {
		status = gatewayv1.ServingStatus_SERVING_STATUS_SERVING
	}
	return &gatewayv1.HealthReport{Status: status}
}

// declaredContracts returns the contracts in a stable order, so two runs of the
// same service send byte-identical registrations.
func (service *Service) declaredContracts() []*gatewayv1.ToolContract {
	service.mutex.RLock()
	defer service.mutex.RUnlock()
	names := make([]string, 0, len(service.contracts))
	for name := range service.contracts {
		names = append(names, name)
	}
	sort.Strings(names)
	contracts := make([]*gatewayv1.ToolContract, 0, len(names))
	for _, name := range names {
		contracts = append(contracts, service.contracts[name])
	}
	return contracts
}

func (service *Service) registryClient() (gatewayv1.RegistryServiceClient, error) {
	service.clientMutex.Lock()
	defer service.clientMutex.Unlock()
	if service.client != nil {
		return service.client, nil
	}
	connection, err := grpc.NewClient(service.config.GatewayAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, fmt.Errorf("sdk: dial gateway %s: %w", service.config.GatewayAddress, err)
	}
	service.connection = connection
	service.client = gatewayv1.NewRegistryServiceClient(connection)
	return service.client, nil
}

// Deadline is a convenience for handlers that want the attempt's deadline as a
// protobuf timestamp, for example to forward it to a downstream system.
func Deadline(instant time.Time) *timestamppb.Timestamp {
	return timestamppb.New(instant)
}
